contracts/btc-finality/src/finality.rs

```diff
@@ -252,29 +252,14 @@ pub fn handle_finality_signature(
     //     this means Babylon will lose safety under an adaptive adversary corrupting even 1
     //     finality provider. It can simply corrupt a new finality provider and equivocate a
     //     historical block over and over again, making a previous block not finalisable forever
-    if fp.slashed_height > 0 && fp.slashed_height < height {
+    if fp.is_slashed() {
         return Err(ContractError::FinalityProviderAlreadySlashed(
             fp_btc_pk_hex.to_string(),
         ));
     }
 
     // Ensure the finality provider has voting power at this height
-    let fp: FinalityProviderInfo = deps
-        .querier
-        .query_wasm_smart(
-            staking_addr.clone(),
-            &btc_staking::msg::QueryMsg::FinalityProviderInfo {
-                btc_pk_hex: fp_btc_pk_hex.to_string(),
-                height: Some(height),
-            },
-        )
-        .map_err(|_| ContractError::NoVotingPower(fp_btc_pk_hex.to_string(), height))?;
-    if fp.power == 0 {
-        return Err(ContractError::NoVotingPower(
-            fp_btc_pk_hex.to_string(),
-            height,
-        ));
-    }
+    ensure_fp_has_power(deps.storage, height, fp_btc_pk_hex)?;
```

Comment:
@gitferry After this change, the finality sig handler is using voting power table in BTC finality contract, not BTC staking contract. It filters out FPs without timestamped pub rand, thus fixes #156

Reply:
This is the important fix.

This could have been done by checking that the FP is in the active set (i.e. in `FP_SET`) at the given height).

I agree this impl is better, as loading the entire set at each finality sig submission would be inefficient.